openmmtools/forces.py

```diff
@@ -321,10 +369,7 @@ def restrained_atom_indices2(self):
     @property
     def restraint_parameters(self):
         """OrderedDict: The restraint parameters in dictionary form."""
-        parameter_values = self.getBondParameters(0)[-1]
-        restraint_parameters = [(self.getPerBondParameterName(parameter_idx), parameter_value)
-                                for parameter_idx, parameter_value in enumerate(parameter_values)]
-        return collections.OrderedDict(restraint_parameters)
+        return collections.OrderedDict(self._restraint_parameters)
```

Comment:
What if we did not rely on the `OpenMM.XmlSerializer` alone to serialize and de-serialize the force? Could we add a `__getstate__` and `__setstate__` for the base force objects and use the OpenMMTools `serialize` and `deserialize` functions?

Reply:
I doubt it is possible unfortunately because I think the serialization of the `System` is performed at the C/C++ level and it automatically serializes the forces in the system in XML format.

Reply:
> The other option would be to infer it from the `energy_function` as you said then?

Yes, for `_controlling_parameter_name`, although the `controlling_parameter_name + ' * (' + energy_function + ')'` is added in the subclass right now, so we may need to push that line up one level in the hierarchy to keep things clean. I believe that line is shared in all subclasses anyway.

> but I am worried that will only lead to confusing namespace overwrites

I agree. I'm also worried about the fact that a serialized system with this method won't be able to be read as a normal OpenMM `System` with `XmlSerializer.deserializeSystem()` without our code.

Reply:
Whops! Sorry I was overthinking it. We'll just have to return `self.getGlobalParameterName(0)` inside the `controlling_parameter_name` property. Maybe also placing an `assert` statement on `__init__()` to check that at that point there are no other global parameters to make sure the index is correct.

Reply:
@andrrizzi on what we talked about, there actually is a `restraint_parameters` property which already exists in the released version. Without some way to generically parse it, I don't know how to preserve this property.

Reply:
Ah, you're right, I forgot about that! Let's just push the implementation of that property in the subclasses then since they know which one to retrieve.

Reply:
Although I don't think anything actually uses that property in OpenMMTools or YANK